File: algorithms/sort/merge_sort/python/merge_sort.py
def merge_sort(li):
    # recursively split the lists
    if len(li) <= 1:
        return li
    middle = len(li) / 2
    left = li[:middle]
    right = li[middle:]
    return merge(merge_sort(left), merge_sort(right))


def merge(left, right):
    # Merge the two lists back together
    result = []
    while left or right:
        if left and right:
            if left[0] < right[0]:
                result.append(left.pop(0))
            else:
                result.append(right.pop(0))
        elif left:
            result.append(left.pop(0))
        else:
            result.append(right.pop(0))
    return result

File: algorithms/sort/merge_sort/haskell/merge_sort.hs
module MergeSort where

{-
Merge Sort in Haskell.
-}

merge :: Ord a =>[a] -> [a] -> [a]
merge [] ys = ys
merge xs [] = xs
merge (x:xs) (y:ys)
  | x > y = y:x:(merge xs ys)
  | y >= x = x:y:(merge xs ys)

mergeSort' :: Ord a => Int -> [a] -> [a]
mergeSort' 1 [x]  = [x]
mergeSort' len xs = merge (mergeSort' middle left) (mergeSort' (len-middle) right)
  where
    middle = len `div` 2
    (left, right) = splitAt middle xs
    
mergeSort :: Ord a => [a] -> [a]
mergeSort [] = []
mergeSort [x] = [x]
mergeSort xs  = mergeSort' (length xs) xs
